rust/pattern/variable/builder/thing_variable_builder.rs

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+use crate::pattern::*;
+use std::fmt::Debug;
+
+pub trait ThingVariableBuilder: Sized {
+    fn has<T: TryInto<Value>>(self, type_name: impl Into<String>, value: T) -> BoundVariable
```

Comment:
All variable builder methods return type-erased `BoundVariable`s. The reason is that we want to be able to pass an array of both `Thing-` and `TypeVariable`s in an array to, e.g., `typeql_match()`, and they therefore have to have the same type.
They would probably have to be erased further into `Pattern`s with introduction of `Disjunction`s and `Negation`s.